Add support for APG (adaptive projected guidance) + unconditionnal SLG

stduhpf opened the pull request:
Implements this paper: https://arxiv.org/abs/2410.02416

### TLDR: 
APG is a set of 3 modilfications for CFG:
- reverse momentum: The CFG delta is getting steered away from (or closer to) the previous step's CFG delta ( `--apg-momentum`)
- normalization: the L2 norm of the CFG delta is clamped to some value "norm threshold" (`--apg-nt`)
- projection: the CFG delta (`out_uncond-out_cond`) is orthogonally projected on the same "direction" as `out_cond`. The final delta is linearly interpolated between the original delta and the projected delta with the parameter "eta" (`--apg-eta`)

Then the guidance update is computed like in normal CFG: `output = out_cond + (cfg_scale-1)*delta`

No extra forward pass is required, so the performance cost is negligible.

Thanks mostly to the normalization, but also the projection, this allows to take adventage of very large CFG scales without getting deep-fried output images. I'm not sure how usefull the reverse momentum really is, but it was in the paper so I added it too (I think it prevents the CFG from going too much "in the same direction" at every step?).

### Usage

`[your usual command with cfg here] --apg-eta 0 --apg-nt 5 --apg-momentum -0.5`  

Recommanded values:
- eta: between 0 and 1, closer to 0 seems better. In the paper, they recommend setting it to 0 altogether (supports any real values though including negatives. Setting it to 1 neutralizes the effect)
- norm: threshold between 1 and 25 depending on the model/prompts (setting it to zero or negative disables the thresholding)
- momentum: preferably negative, ideally between 0 and -1 (again, any value is technically supported, setting it to 0 neutralizes the effect)

Feel free to play around with the settings, going outside of the recommended ranges can have interesting effects, especially with eta and momentum.

### Tips

To help you figure out the right setting for the norm threshold, you can use the `SD_LOG_CFG_DELTA_NORM` environement variable to `"ON"` (for example, on windows powershell: `$env:SD_LOG_CFG_DELTA_NORM="ON"`). Then you can run your model with normal CFG. 

This will print a number at each step on the terminal/logger (coresponding to the unclamped L2 norm of the CFG delta).

Pick a number that's within the range of values printed as the base norm threshold (like the median for example). 

If you want to use CFG scale that's above the recomended one for the model you're using, I recommend using something this formula to update the threshold accordingly:

`apg-nt = base_norm_threshold*(recomended_cfg_scale-1)/(cfg_scale-1)`

The ideal parameters will depending on the prompt and other settings, but they will most likely stay in the same order of magnitude for the same model.

***

I also added an experimental smoothing parameter (`--apg-nt-smoothing`) for the normalization. In the paper they're using a "saturate" function (`min(1,threshold/norm)`), which has two potential issues: it has a kink (not continuously differentiable), and is not invertible as all input values outside of the $[0,1]$ range get mapped to $1$.

 This experimental feature remplaces the $min(1,x)$ function with $\frac{x}{\left(1+x^{\frac{1}{p}}\right)^{p}}$, which is smooth and invertible. It is equivalent to $f(x)=x$ for small values of $x$ (just like the min) and perfectly approximates to the original $min(1,x)$ as the value of $p$ goes to $0$.
 
A good value of the smoothing parameter would map the upper bound of the CFG delta norms to somewhere in the [0.95, 0.99] range with this formula. There is no closed form formula to find a good value, but you can just try things and see how it goes.
 
 I made this to experiment with the values for picking the threshold parameters: https://www.desmos.com/calculator/7sir5unorl
 
 *** 
 Edit: I also added unconditionnal SLG (`--slg-uncond`) (I stole the idea from https://github.com/deepbeepmeep/Wan2GP/pull/61)
 
 Just a simpler version of SLG (Skip Layer Guidance, introduced in #451) for DiT models.
 
Default SLG requires a third forward pass of the network with some layers skipped. This increase the computing time by a bit under 50% for the SLG steps, wich isn't ideal. 

Unconditionnal SLG skips layers during the same unconditionnal pass used for CFG/APG. It seems to be about as effective as normal SLG, but it's even faster than CFG, thanks to the layers being skipped.

Downside: it's less flexible, `--slg-scale` should be kept to 0 and `--cfg-scale` now controls both the CFG and the SLG. 
Upside: It's faster.

setting both `--slg-scale != 0` and `--slg-uncond` at the same time will most likely degrade image quality while using more compute. It's possible, but not recommended. (Maybe it could be worth to investigate skipping a different sets of layers with normal slg and unconditionnal slg, but we're getting too far out of scope for this PR)

stduhpf committed:
apg: first implementation

refactor guidance params in lib

main: add apg support

add apg settings to image params

Fix cfg 1 crash

Fix CI build

stduhpf committed:
apg: add experimental threshold smoothing parameter

stduhpf committed:
add uncond slg variant

fix default slg params

wbruna commented:
APG works even with distilled models. I was able to get good LCM generations with 4+ CFG, and _negatives_.

stduhpf committed:
apg: add SD_LOG_CFG_DELTA_NORM

stduhpf committed:
fix apg cli args with new api

stduhpf committed:
integrate apg into new API

stduhpf commented:
I tried to make it work with txt+img guidance, it's a bit of a hack, but it seems to work decently well.

wbruna committed:
merged master fd693ac6a2ab into apg (#7)

* docs: add sd.cpp-webui as an available frontend (#738)

* fix: correct head dim check and L_k padding of flash attention (#736)

* fix: convert f64 to f32 and i64 to i32 when loading weights

* docs: add LocalAI to README's UIs (#741)

* sync: update ggml

* sync: update ggml

* feat: upgrade musa sdk to rc4.2.0 (#732)

* feat: change image dimensions requirement for DiT models (#742)

* feat: add missing models and parameters to image metadata (#743)

* feat: add new scheduler types, clip skip and vae to image embedded params

- If a non default scheduler is set, include it in the 'Sampler' tag in the data
embedded into the final image.
- If a custom VAE path is set, include the vae name (without path and extension)
in embedded image params under a `VAE:` tag.
- If a custom Clip skip is set, include that Clip skip value in embedded image
params under a `Clip skip:` tag.

* feat: add separate diffusion and text models to metadata

---------

Co-authored-by: one-lithe-rune <skapusniak@lithe-runes.com>

* refector: optimize the usage of tensor_types

* feat: support build against system installed GGML library (#749)

* chore: avoid setting GGML_MAX_NAME when building against external ggml (#751)

An external ggml will most likely have been built with the default
GGML_MAX_NAME value (64), which would be inconsistent with the value
set by our build (128). That would be an ODR violation, and it could
easily cause memory corruption issues due to the different
sizeof(struct ggml_tensor) values.

For now, when linking against an external ggml, we demand it has been
patched with a bigger GGML_MAX_NAME, since we can't check against a
value defined only at build time.

* Conv2D direct support (#744)

* Conv2DDirect for VAE stage

* Enable only for Vulkan, reduced duplicated code

* Cmake option to use conv2d direct

* conv2d direct always on for opencl

* conv direct as a flag

* fix merge typo

* Align conv2d behavior to flash attention's

* fix readme

* add conv2d direct for controlnet

* add conv2d direct for esrgan

* clean code, use enable_conv2d_direct/get_all_blocks

* format code

---------

Co-authored-by: leejet <leejet714@gmail.com>

* sync: update ggml, make cuda im2col a little faster

* chore: add Nvidia 30 series (cuda arch 86) to build

* feat: throttle model loading progress updates (#782)

Some terminals have slow display latency, so frequent output
during model loading can actually slow down the process.

Also, since tensor loading times can vary a lot, the progress
display now shows the average across past iterations instead
of just the last one.

* docs: add missing dash to docs/chroma.md (#771)

* docs: add compile option needed by Ninja (#770)

* feat: show usage on unknown arg (#767)

* fix: typo in the verbose long flag (#783)

* feat: add wan2.1/2.2 support (#778)

* add wan vae suppport

* add wan model support

* add umt5 support

* add wan2.1 t2i support

* make flash attn work with wan

* make wan a little faster

* add wan2.1 t2v support

* add wan gguf support

* add offload params to cpu support

* add wan2.1 i2v support

* crop image before resize

* set default fps to 16

* add diff lora support

* fix wan2.1 i2v

* introduce sd_sample_params_t

* add wan2.2 t2v support

* add wan2.2 14B i2v support

* add wan2.2 ti2v support

* add high noise lora support

* sync: update ggml submodule url

* avoid build failure on linux

* avoid build failure

* update ggml

* update ggml

* fix sd_version_is_wan

* update ggml, fix cpu im2col_3d

* fix ggml_nn_attention_ext mask

* add cache support to ggml runner

* fix the issue of illegal memory access

* unify image loading processing

* add wan2.1/2.2 FLF2V support

* fix end_image mask

* update to latest ggml

* add GGUFReader

* update docs

* feat: add support for timestep boundary based automatic expert routing in Wan MoE (#779)

* Wan MoE: Automatic expert routing based on timestep boundary

* unify code style and fix some issues

---------

Co-authored-by: leejet <leejet714@gmail.com>

* feat: add flow shift parameter (for SD3 and Wan) (#780)

* Add flow shift parameter (for SD3 and Wan)

* unify code style and fix some issues

---------

Co-authored-by: leejet <leejet714@gmail.com>

* docs: update docs and help message

* chore: update to c++17

* docs: update docs/wan.md

* fix: add flash attn support check (#803)

* feat: support incrementing ref image index (omni-kontext) (#755)

* kontext: support  ref images indices

* lora: support x_embedder

* update help message

* Support for negative indices

* support for OmniControl (offsets at index 0)

* c++11 compat

* add --increase-ref-index option

* simplify the logic and fix some issues

* update README.md

* remove unused variable

---------

Co-authored-by: leejet <leejet714@gmail.com>

* feat: add detailed tensor loading time stat (#793)

* fix: clarify lora quant support and small fixes (#792)

* fix: accept NULL in sd_img_gen_params_t::input_id_images_path (#809)

* chore: update flash attention warnings (#805)

* fix: use {} for params init instead of memset (#781)

* chore: remove sd3 flash attention warn (#812)

* feat: use log_printf to print ggml logs (#545)

* chore: add install() support in CMakeLists.txt (#540)

* feat: add SmoothStep Scheduler (#813)

* feat: add sd3 flash attn support (#815)

* fix: make tiled VAE reuse the compute buffer (#821)

* feat: reduce CLIP memory usage with no embeddings (#768)

* fix: make weight override more robust against ggml changes (#760)

* fix: do not force VAE type to f32 on SDXL (#716)

This seems to be a leftover from the initial SDXL support: it's
not enough to avoid NaN issues, and it's not not needed for the
fixed sdxl-vae-fp16-fix .

* feat: use Euler sampling by default for SD3 and Flux (#753)

Thank you for your contribution.

* fix: harden for large files (#643)

* feat: Add SYCL Dockerfile (#651)

* feat: increase work_ctx memory buffer size (#814)

* docs: update docs

* feat: add VAE encoding tiling support and adaptive overlap  (#484)

* implement  tiling vae encode support

* Tiling (vae/upscale): adaptative overlap

* Tiling: fix edge case

* Tiling: fix crash when less than 2 tiles per dim

* remove extra dot

* Tiling: fix edge cases for adaptative overlap

* tiling: fix edge case

* set vae tile size via env var

* vae tiling: refactor again, base on smaller buffer for alignment

* Use bigger tiles for encode (to match compute buffer size)

* Fix edge case when tile is bigger than latent

* non-square VAE tiling (#3)

* refactor tile number calculation

* support non-square tiles

* add env var to change tile overlap

* add safeguards and better error messages for SD_TILE_OVERLAP

* add safeguards and include overlapping factor for SD_TILE_SIZE

* avoid rounding issues when specifying SD_TILE_SIZE as a factor

* lower SD_TILE_OVERLAP limit

* zero-init empty output buffer

* Fix decode latent size

* fix encode

* tile size params instead of env

* Tiled vae parameter validation (#6)

* avoid crash with invalid tile sizes, use 0 for default

* refactor default tile size, limit overlap factor

* remove explicit parameter for relative tile size

* limit encoding tile to latent size

* unify code style and format code

* update docs

* fix get_tile_sizes in decode_first_stage

---------

Co-authored-by: Wagner Bruna <wbruna@users.noreply.github.com>
Co-authored-by: leejet <leejet714@gmail.com>

* feat: add vace support (#819)

* add wan vace t2v support

* add --vace-strength option

* add vace i2v support

* fix the processing of vace_context

* add vace v2v support

* update docs

* feat: optimize tensor loading time (#790)

* opt tensor loading

* fix build failure

* revert the changes

* allow the use of n_threads

* fix lora loading

* optimize lora loading

* add mutex

* use atomic

* fix build

* fix potential duplicate issue

* avoid duplicate lookup of lora tensor

* fix progeress bar

* remove unused remove_duplicates

---------

Co-authored-by: leejet <leejet714@gmail.com>

* refactor: simplify the logic of pm id image loading (#827)

* feat: add sgm_uniform scheduler, simple scheduler, and support for NitroFusion (#675)

* feat: Add timestep shift and two new schedulers

* update readme

* fix spaces

* format code

* simplify SGMUniformSchedule

* simplify shifted_timestep logic

* avoid conflict

---------

Co-authored-by: leejet <leejet714@gmail.com>

* refactor: move tiling cacl and debug print into the tiling code branch (#833)

* refactor: simplify DPM++ (2S) Ancestral (#667)

* chore: set release tag by commit count

* chore: fix workflow (#836)

* fix: avoid multithreading issues in the model loader

* fix: avoid segfault for pix2pix models without reference images (#766)

* fix: avoid segfault for pix2pix models with no reference images

* fix: default to empty reference on pix2pix models to avoid segfault

* use resize instead of reserve

* format code

---------

Co-authored-by: leejet <leejet714@gmail.com>

* refactor: remove unused --normalize-input parameter (#835)

* fix: correct tensor deduplication logic (#844)

* docs: include Vulkan compatibility for LoRA quants (#845)

* docs: HipBLAS / ROCm build instruction fix (#843)

* fix: tensor loading thread count (#854)

* fix: optimize the handling of CLIP embedding weight (#840)

* sync: update ggml

* sync: update ggml

* fix: optimize the handling of embedding weight (#859)

* feat: add support for Flux Controls and Flex.2 (#692)

* docs: update README.md (#866)

* chore: fix dockerfile libgomp1 dependency + improvements (#852)

* fix: ensure directory iteration results are sorted by filename (#858)

* chore: fix vulkan ci (#878)

* feat: add support for more esrgan models & x2 & x1 models (#855)

* feat: add a stand-alone upscale mode (#865)

* feat: add a stand-alone upscale mode

* fix prompt option check

* format code

* update README.md

---------

Co-authored-by: leejet <leejet714@gmail.com>

* refactor: deal with default img-cfg-scale at the library level (#869)

* feat: add Qwen Image support (#851)

* add qwen tokenizer

* add qwen2.5 vl support

* mv qwen.hpp -> qwenvl.hpp

* add qwen image model

* add qwen image t2i pipeline

* fix qwen image flash attn

* add qwen image i2i pipline

* change encoding of vocab_qwen.hpp to utf8

* fix get_first_stage_encoding

* apply jeffbolz f32 patch

https://github.com/leejet/stable-diffusion.cpp/pull/851#issuecomment-3335515302

* fix the issue that occurs when using CUDA with k-quants weights

* optimize the handling of the FeedForward precision fix

* to_add_out precision fix

* update docs

* fix: resolve VAE tiling problem in Qwen Image (#873)

* fix: avoid generating black images when running T5 on the GPU (#882)

* fix: correct canny preprocessor (#861)

* fix: better progress display for second-order samplers (#834)

* feat: add Qwen Image Edit support (#877)

* add ref latent support for qwen image

* optimize clip_preprocess and fix get_first_stage_encoding

* add qwen2vl vit support

* add qwen image edit support

* fix qwen image edit pipeline

* add mmproj file support

* support dynamic number of Qwen image transformer blocks

* set prompt_template_encode_start_idx every time

* to_add_out precision fix

* to_out.0 precision fix

* update docs

---------

Co-authored-by: Daniele <57776841+daniandtheweb@users.noreply.github.com>
Co-authored-by: Erik Scholz <Green-Sky@users.noreply.github.com>
Co-authored-by: leejet <leejet714@gmail.com>
Co-authored-by: Ettore Di Giacinto <mudler@users.noreply.github.com>
Co-authored-by: R0CKSTAR <xiaodong.ye@mthreads.com>
Co-authored-by: stduhpf <stephduh@live.fr>
Co-authored-by: one-lithe-rune <skapusniak@lithe-runes.com>
Co-authored-by: Seas0 <seashkey@gmail.com>
Co-authored-by: NekopenDev <197017459+nekopendev@users.noreply.github.com>
Co-authored-by: SmallAndSoft <45131567+SmallAndSoft@users.noreply.github.com>
Co-authored-by: Markus Hartung <mail@hartmark.se>
Co-authored-by: clibdev <52199778+clibdev@users.noreply.github.com>
Co-authored-by: Richard Palethorpe <io@richiejp.com>
Co-authored-by: rmatif <kingrealriadh@gmail.com>
Co-authored-by: vmobilis <75476228+vmobilis@users.noreply.github.com>
Co-authored-by: Stefan-Olt <stefan-oltmanns@gmx.net>
Co-authored-by: Sharuzzaman Ahmat Raslan <sharuzzaman@gmail.com>
Co-authored-by: Serkan Sahin <14278530+SergeantSerk@users.noreply.github.com>
Co-authored-by: Pedrito <pedro.c.vfx@gmail.com>

leejet committed:
Merge branch 'master' into apg

leejet committed:
remove cfg delta norm log
